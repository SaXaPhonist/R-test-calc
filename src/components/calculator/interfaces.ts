export interface ICalcButton {
  value: string;
  type: 'operation' | 'number' | 'syntx' | 'clear';
  priority?: number;
}

export interface IMathOperations {
  [key: string]: (val1: number, val2: number) => number;
  '\u221A': (val1: number) => number;
}

export interface IButton {
  value: string;
  type: 'number' | 'operation' | 'exec' | 'braket' | 'divider' | 'clear';
  priority?: number;
}

export interface IButtonNumber extends Omit<IButton, 'type' | 'priority'> {
  type: 'number';
}

export interface IButtonOperation extends Omit<IButton, 'type' | 'priority'> {
  type: 'operation';
  priority: number;
}

export interface IButtonBraket extends Omit<IButton, 'type' | 'priority'> {
  type: 'braket';
}
