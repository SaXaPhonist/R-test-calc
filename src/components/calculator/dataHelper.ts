import { IMathOperations } from './interfaces';

export const buttonsLibrary = [
  { value: 'C', type: 'clear' },
  { value: '\u221A', type: 'operation', priority: 3 },
  { value: '%', type: 'operation', priority: 2 },
  { value: '/', type: 'operation', priority: 2 },
  { value: '1', type: 'number' },
  { value: '2', type: 'number' },
  { value: '3', type: 'number' },
  { value: '*', type: 'operation', priority: 2 },
  { value: '4', type: 'number' },
  { value: '5', type: 'number' },
  { value: '6', type: 'number' },
  { value: '-', type: 'operation', priority: 1 },
  { value: '7', type: 'number' },
  { value: '8', type: 'number' },
  { value: '9', type: 'number' },
  { value: '+', type: 'operation', priority: 1 },
  { value: '(', type: 'braket' },
  { value: ')', type: 'braket' },
  { value: '0', type: 'number' },
  { value: '.', type: 'devider' },
  { value: '=', type: 'exec' },
];

export const mathOperations: IMathOperations = {
  '+': (val1, val2) => val1 + val2,
  '-': (val1, val2) => val1 - val2,
  '*': (val1, val2) => val1 * val2,
  '/': (val1, val2) => val1 / val2,
  '\u221A': (val1) => Math.sqrt(val1),
  '%': (val1, val2) => (val1 * val2) / 100,
};

export const keyMapper: Record<string, string> = {
  '1': '1',
  '2': '2',
  '3': '3',
  '4': '4',
  '5': '5',
  '6': '6',
  '7': '7',
  '8': '8',
  '9': '9',
  '0': '0',
  '.': '.',
  '=': '=',
  Enter: '=',
  '+': '+',
  '-': '-',
  '*': '*',
  '/': '/',
  '%': '%',
  Escape: 'C',
  Clear: 'C',
  Backspace: 'C',
  '^': '\u221A',
  '(': '(',
  ')': ')',
};
