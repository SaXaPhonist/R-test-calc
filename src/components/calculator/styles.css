.calculator-wrapper {
    margin: 2em auto;
    padding: 1.5em ;
    width: 464px;
    background-color: rgba(256, 256, 256, 0.2);
    border-radius: 5%;
}

.calculator {
    border-radius: 5%;
    width: 100%;
    padding: 3em 1em 1em ;
    box-shadow: 3px 10px 10px 9px rgba(34, 60, 80, 0.2);
    background: linear-gradient(342deg, rgba(105,122,196,1) 0%, rgba(72,92,177,1) 69%, rgba(59,76,148,1) 100%);
}
.calculator__presentation-box{
    height: 2em;
    color: #fff;
}

.display {
    display: flex;
    align-items: flex-end;
    flex-direction: column;
    margin-bottom: .5em;
    border: none;
	border-bottom: 0.125rem solid rgb(179, 176, 176);
}

.calculator__input-container {
    height: 2.2em;
	font-size: 1.5rem;
	padding-left: 0.875rem;
	line-height: 147.6%;
	padding-top: 0.825rem;
    background-color: transparent;
    color: #fff;
}

.calculator__input-container:focus {
    outline: none;
}


.calculator__buttons-container {
    display: flex;
    flex-wrap: wrap;
    flex-direction: row;
    justify-content: center;
}


.calculator__button {
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: transparent;
    border: none;
    border-radius: 50%;
    color: #fff;
    width: 6em;
    height: 6em;
}

.active {
    background-color: rgba(147, 251, 255, 0.302);
}

.calculator__button > span {
    font-size: 2rem
}

.calculator__button:hover {
    background-color: rgb(135, 135, 224);
    cursor: pointer;
}