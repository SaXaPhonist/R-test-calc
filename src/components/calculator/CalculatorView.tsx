import { ICalcButton } from './interfaces';
import './styles.css';

export const buttonsLibrary: ICalcButton[] = [
  { value: 'C', type: 'clear' },
  { value: '\u221A', type: 'operation', priority: 3 },
  { value: '%', type: 'operation', priority: 2 },
  { value: '/', type: 'operation', priority: 2 },
  { value: '1', type: 'number' },
  { value: '2', type: 'number' },
  { value: '3', type: 'number' },
  { value: '*', type: 'operation', priority: 2 },
  { value: '4', type: 'number' },
  { value: '5', type: 'number' },
  { value: '6', type: 'number' },
  { value: '-', type: 'operation', priority: 1 },
  { value: '7', type: 'number' },
  { value: '8', type: 'number' },
  { value: '9', type: 'number' },
  { value: '+', type: 'operation', priority: 1 },
  { value: '(', type: 'syntx', priority: 4 },
  { value: ')', type: 'syntx', priority: 4 },
  { value: '0', type: 'number' },
  { value: '.', type: 'number' },
  { value: '=', type: 'operation' },
];

interface IProps {
  calcValue: string;
  handleClick: (e: React.MouseEvent) => void;
  handleKeyPress: () => void;
  result: string | undefined;
}

export const Calculator = ({ calcValue, result, handleClick, handleKeyPress }: IProps) => {
  return (
    <div className="calculator-wrapper">
      <div className="calculator">
        <section className="calculator__presentation-box">{result || calcValue}</section>
        <input value={calcValue} className="calculator__input-container" />
        <section className="calculator__buttons-container">
          {buttonsLibrary.map((button) => (
            <div
              key={button.value}
              role="none"
              onFocus={() => {}}
              onKeyDown={handleKeyPress}
              onClick={handleClick}
              className="calculator__button"
            >
              <p>{button.value}</p>
            </div>
          ))}
        </section>
      </div>
    </div>
  );
};
