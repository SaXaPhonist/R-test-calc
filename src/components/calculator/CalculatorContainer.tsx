/* eslint-disable no-plusplus */
/* eslint-disable consistent-return */
/* eslint-disable no-param-reassign */

import { useCallback, useLayoutEffect, useState } from 'react';
import { buttonsLibrary, Calculator } from './CalculatorView';

interface IMathOperations {
  [key: string]: (val1: number, val2: number) => number;
  '\u221A': (val1: number) => number;
}

const mathOperations: IMathOperations = {
  '+': (val1, val2) => val1 + val2,
  '-': (val1, val2) => val1 - val2,
  '*': (val1, val2) => val1 * val2,
  '/': (val1, val2) => val1 / val2,
  '\u221A': (val1) => Math.sqrt(val1),
  '%': (val1, val2) => (val1 * val2) / 100,
};

export const CalculatorContainer = () => {
  const [maxDepth, setMaxDepth] = useState(0);
  const [inputValue, setInputValue] = useState('');
  const [lastBraket, setLastBraket] = useState('');
  const [openBraketsCount, setOpenBraketsCount] = useState(0);
  const [closeBraketCount, setCloseBraketCount] = useState(0);
  const [lastNumber, setLastNumber] = useState<string>('');
  const [result, setResult] = useState<string | undefined>();
  const [expression, setExpression] = useState<
    Array<{ value: string; type: 'number' | 'operation' | 'syntx'; priority?: number }>
  >([]);
  const [lastVal, setLastVal] = useState<string | undefined>();

  const clearAll = () => {
    setInputValue('');
    setLastNumber('');
    setExpression([]);
    setMaxDepth(0);
    setCloseBraketCount(0);
    setOpenBraketsCount(0);
    setLastBraket('');
    setResult(undefined);
  };

  const addNumber = (newNumber: string) => {
    if (lastNumber) {
      setLastNumber((prev) => `${prev}${newNumber}`);
      setInputValue((prev) => `${prev}${newNumber}`);
      return;
    }
    setLastNumber(newNumber);
    setInputValue((prev) => `${prev}${newNumber}`);
  };

  const mathDepth = (
    expressionArray: Array<{
      value: string;
      priority?: number;
      type: 'number' | 'operation' | 'syntx';
    }>,
    depth: number,
  ) => {
    for (let i = depth; i >= 1; i -= 1) {
      for (let j = 0; j <= expressionArray.length; j += 1) {
        const currentEl = expressionArray[j];
        if (!currentEl) {
          break;
        }
        if (expressionArray.length === 1) {
          return expressionArray[0].value;
        }
        if (currentEl.type === 'syntx' && currentEl.value === '(') {
          const nestedArray = expressionArray.slice(j);
          const closeBraketIndx = nestedArray.findIndex(
            (el) => el.value === ')' && el.priority === currentEl.priority,
          );
          const nestedExpression = nestedArray.slice(1, closeBraketIndx);
          const mathResult = parseFloat(mathDepth(nestedExpression, depth - 1));
          expressionArray.splice(j, nestedExpression.length, {
            value: mathResult.toString(),
            type: 'number',
          });
          j = 0;
        } else if (currentEl.type === 'operation' && currentEl.value === '-' && j === 0) {
          const nextNum = parseFloat(expressionArray[j + 1].value);
          const mathResult = -nextNum;
          expressionArray.splice(j, 2, { value: mathResult.toString(), type: 'number' });
          j = 0;
        } else if (currentEl.type === 'operation' && currentEl.value === '\u221A') {
          const nextNum = parseFloat(expressionArray[j + 1].value);
          const mathResult = mathOperations[currentEl.value](nextNum);
          expressionArray.splice(j, 2, { value: mathResult.toString(), type: 'number' });
          j = 0;
        } else if (currentEl.type === 'operation' && currentEl.priority === i) {
          const prevNum = parseFloat(expressionArray[j - 1].value);
          const nextNum = parseFloat(expressionArray[j + 1].value);
          const mathResult = mathOperations[currentEl.value](prevNum, nextNum);
          expressionArray.splice(j - 1, 3, { value: mathResult.toString(), type: 'number' });
          j = 0;
        }
      }
    }
    return expressionArray[0].value;
  };

  const getReuslt = useCallback(() => {
    const resultCalc = mathDepth(expression, maxDepth);
    if (resultCalc) {
      setResult(resultCalc);
      setInputValue(resultCalc);
      setLastNumber(resultCalc);
    }
    setMaxDepth(0);
    setCloseBraketCount(0);
    setOpenBraketsCount(0);
    setLastBraket('');
  }, [expression, maxDepth]);

  useLayoutEffect(() => {
    if (lastVal && lastVal === '=') {
      getReuslt();
    }
  }, [lastVal, getReuslt]);

  const expressionConstructor = (value?: string) => {
    const [button] = buttonsLibrary.filter((btn) => btn.value === value);

    if (button.type === 'operation') {
      if (lastVal && Object.keys(mathOperations).includes(lastVal)) return;
      if (!lastNumber && button.value !== '\u221A' && button.value !== '-') return;
      if (button.value === '=') {
        if (lastNumber) {
          setExpression((prev) => [...prev, { value: lastNumber, type: 'number' }]);
        }
        return;
      }
      if (button.priority && maxDepth < button.priority) {
        setMaxDepth(button.priority);
      }
      if (lastNumber) {
        setExpression((prev) => [
          ...prev,
          { value: lastNumber, type: 'number' },
          { value: button.value, type: 'operation', priority: button.priority },
        ]);
        setInputValue((prev) => `${prev}${button.value}`);
        setLastNumber('');
        return;
      }
      setExpression((prev) => [
        ...prev,
        { value: button.value, type: 'operation', priority: button.priority },
      ]);
      setInputValue((prev) => `${prev}${button.value}`);
      return;
    }
    if (button.type === 'number') {
      if (button.value === '.' && lastVal === '.') {
        return;
      }
      addNumber(button.value);
    }
    if (button.type === 'syntx' && button.priority) {
      let bracketPriority: number;
      if (lastBraket === '(' && button.value === '(') {
        bracketPriority = maxDepth + 1;
        setMaxDepth(bracketPriority);
        setExpression((prev) => [
          ...prev,
          { value: button.value, type: 'syntx', priority: bracketPriority },
        ]);
        setInputValue((prev) => `${prev}${button.value}`);
        setLastBraket(button.value);
        setOpenBraketsCount((prev) => prev + 1);
      } else if (!lastBraket && button.value === '(') {
        bracketPriority = button.priority;
        setMaxDepth(bracketPriority);
        if (lastNumber) {
          setExpression((prev) => [
            ...prev,
            { value: lastNumber, type: 'number' },
            { value: button.value, type: 'syntx', priority: bracketPriority },
          ]);
          setLastNumber('');
        } else {
          setExpression((prev) => [
            ...prev,
            { value: button.value, type: 'syntx', priority: bracketPriority },
          ]);
        }
        setInputValue((prev) => `${prev}${button.value}`);
        setLastBraket(button.value);
        setOpenBraketsCount((prev) => prev + 1);
      } else if (lastBraket === '(' && button.value === ')') {
        if (lastNumber) {
          setExpression((prev) => [
            ...prev,
            { value: lastNumber, type: 'number' },
            { value: button.value, type: 'syntx', priority: maxDepth },
          ]);
          setLastNumber('');
        } else {
          setExpression((prev) => [
            ...prev,
            { value: button.value, type: 'syntx', priority: bracketPriority },
          ]);
        }
        setInputValue((prev) => `${prev}${button.value}`);
        setLastBraket(button.value);
        setCloseBraketCount((prev) => prev + 1);
      } else if (
        lastBraket === ')' &&
        button.value === ')' &&
        closeBraketCount < openBraketsCount
      ) {
        if (lastNumber) {
          setExpression((prev) => [
            ...prev,
            { value: lastNumber, type: 'number' },
            { value: button.value, type: 'syntx', priority: maxDepth - closeBraketCount },
          ]);
          setLastNumber('');
        } else {
          setExpression((prev) => [
            ...prev,
            { value: button.value, type: 'syntx', priority: bracketPriority },
          ]);
        }
        setInputValue((prev) => `${prev}${button.value}`);
        setLastBraket(button.value);
        setCloseBraketCount((prev) => prev + 1);
      } else {
        throw new Error('Syntax ERROR');
      }
    }
    if (button.type === 'clear') {
      clearAll();
    }
  };

  const extractValue = (e: React.MouseEvent) => {
    const tag = (e.target as HTMLElement).tagName;
    if (tag === 'DIV') {
      const element = e.target as HTMLButtonElement;
      return element.querySelector('p')?.innerText;
    }
    if (tag === 'P') {
      const element = e.target as HTMLParagraphElement;
      return element.innerText;
    }
    return undefined;
  };

  const handleClick = (e: React.MouseEvent) => {
    const value = extractValue(e);
    setLastVal(value);
    expressionConstructor(value);
  };

  const handleKeyPress = () => {};

  return (
    <Calculator
      calcValue={inputValue}
      result={result}
      handleClick={handleClick}
      handleKeyPress={handleKeyPress}
    />
  );
};
