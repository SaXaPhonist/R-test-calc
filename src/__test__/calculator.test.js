/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-call */
import '@testing-library/jest-dom/extend-expect';
import { cleanup, fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { CalculatorContainer } from '../components/calculator/CalculatorContainer';

describe('Calculator component', () => {
  let container;
  beforeEach(() => {
    container = render(<CalculatorContainer />);
  });

  afterEach(cleanup);

  it('Calculator match snapshot', () => {
    expect(container).toMatchSnapshot();
  });

  it('+ operation', async () => {
    const user = userEvent.setup();
    const button1 = container.getByTestId('calculator-button-1');
    const button2 = container.getByTestId('calculator-button-2');
    const buttonPlus = container.getByTestId('calculator-button-+');
    const buttonExec = container.getByTestId('calculator-button-=');

    user.click(button1);
    user.click(buttonPlus);
    user.click(button2);
    user.click(buttonExec);

    const answerBox = await container.findByTestId('calculator-answer');
    await waitFor(() => expect(answerBox.textContent).toBe('3'));
  });

  it('- operation', async () => {
    const user = userEvent.setup();
    const button1 = container.getByTestId('calculator-button-5');
    const button2 = container.getByTestId('calculator-button-3');
    const buttonPlus = container.getByTestId('calculator-button--');
    const buttonExec = container.getByTestId('calculator-button-=');

    user.click(button1);
    user.click(buttonPlus);
    user.click(button2);
    user.click(buttonExec);

    const answerBox = await container.findByTestId('calculator-answer');
    await waitFor(() => expect(answerBox.textContent).toBe('2'));
  });

  it('* operation', async () => {
    const user = userEvent.setup();
    const button1 = container.getByTestId('calculator-button-5');
    const button2 = container.getByTestId('calculator-button-3');
    const buttonPlus = container.getByTestId('calculator-button-*');
    const buttonExec = container.getByTestId('calculator-button-=');

    user.click(button1);
    user.click(buttonPlus);
    user.click(button2);
    user.click(buttonExec);

    const answerBox = await container.findByTestId('calculator-answer');
    await waitFor(() => expect(answerBox.textContent).toBe('15'));
  });

  it('/ operation', async () => {
    const user = userEvent.setup();
    const button1 = container.getByTestId('calculator-button-1');
    const button6 = container.getByTestId('calculator-button-6');
    const button2 = container.getByTestId('calculator-button-4');
    const buttonPlus = container.getByTestId('calculator-button-/');
    const buttonExec = container.getByTestId('calculator-button-=');

    user.click(button1);
    user.click(button6);
    user.click(buttonPlus);
    user.click(button2);
    user.click(buttonExec);

    const answerBox = await container.findByTestId('calculator-answer');
    await waitFor(() => expect(answerBox.textContent).toBe('4'));
  });

  it('% operation', async () => {
    const user = userEvent.setup();
    const button2 = container.getByTestId('calculator-button-2');
    const button0 = container.getByTestId('calculator-button-0');
    const button8 = container.getByTestId('calculator-button-8');
    const opeartionBtn = container.getByTestId('calculator-button-%');
    const buttonExec = container.getByTestId('calculator-button-=');

    user.click(button2);
    user.click(button0);
    user.click(opeartionBtn);
    user.click(button8);
    user.click(button0);
    user.click(buttonExec);

    const answerBox = await container.findByTestId('calculator-answer');
    await waitFor(() => expect(answerBox.textContent).toBe('16'));
  });

  it('sqrt operation', async () => {
    const user = userEvent.setup();
    const button6 = container.getByTestId('calculator-button-6');
    const button4 = container.getByTestId('calculator-button-4');
    const opeartionBtn = container.getByTestId('calculator-button-\u221A');
    const buttonExec = container.getByTestId('calculator-button-=');
    user.click(button6);
    user.click(button4);
    user.click(opeartionBtn);
    user.click(buttonExec);

    const answerBox = await container.findByTestId('calculator-answer');
    await waitFor(() => expect(answerBox.textContent).toBe('8'));
  });

  it('clear all operation', async () => {
    const user = userEvent.setup();
    const button6 = container.getByTestId('calculator-button-6');
    const button4 = container.getByTestId('calculator-button-4');
    const buttonClear = container.getByTestId('calculator-button-C');
    const opeartionBtn = container.getByTestId('calculator-button-+');
    user.click(button6);
    user.click(opeartionBtn);
    user.click(button4);
    user.click(buttonClear);
    const answerBox = await container.findByTestId('calculator-answer');
    await waitFor(() => expect(answerBox.textContent).toBe(''));
  });

  it('float number operation', async () => {
    const user = userEvent.setup();
    const button6 = container.getByTestId('calculator-button-6');
    const button4 = container.getByTestId('calculator-button-4');
    const button0 = container.getByTestId('calculator-button-0');
    const divider = container.getByTestId('calculator-button-.');
    const buttonExec = container.getByTestId('calculator-button-=');
    const opeartionBtn = container.getByTestId('calculator-button-+');
    user.click(button0);
    user.click(divider);
    user.click(button6);
    user.click(opeartionBtn);
    user.click(button0);
    user.click(divider);
    user.click(button4);
    user.click(buttonExec);
    const answerBox = await container.findByTestId('calculator-answer');
    await waitFor(() => expect(answerBox.textContent).toBe('1'));
  });

  it('operation with brakets', async () => {
    const user = userEvent.setup();
    const button6 = container.getByTestId('calculator-button-6');
    const button4 = container.getByTestId('calculator-button-4');
    const braketOpen = container.getByTestId('calculator-button-(');
    const braketClose = container.getByTestId('calculator-button-)');
    const buttonExec = container.getByTestId('calculator-button-=');
    const opeartionPlus = container.getByTestId('calculator-button-+');
    const opeartionMinus = container.getByTestId('calculator-button--');

    user.click(button6);
    user.click(opeartionPlus);
    user.click(braketOpen);
    user.click(button6);
    user.click(opeartionMinus);
    user.click(button4);
    user.click(braketClose);
    user.click(buttonExec);

    const answerBox = await container.findByTestId('calculator-answer');
    await waitFor(() => expect(answerBox.textContent).toBe('8'));
  });

  it('priority operations', async () => {
    const user = userEvent.setup();
    const button6 = container.getByTestId('calculator-button-6');
    const button4 = container.getByTestId('calculator-button-4');
    const buttonExec = container.getByTestId('calculator-button-=');
    const opeartionPlus = container.getByTestId('calculator-button-+');
    const opeartionMinus = container.getByTestId('calculator-button--');
    const opeartionMultipe = container.getByTestId('calculator-button-*');

    user.click(button6);
    user.click(opeartionPlus);
    user.click(button6);
    user.click(opeartionMultipe);
    user.click(button4);
    user.click(opeartionMinus);
    user.click(button4);
    user.click(buttonExec);

    const answerBox = await container.findByTestId('calculator-answer');
    await waitFor(() => expect(answerBox.textContent).toBe('26'));
  });
});
