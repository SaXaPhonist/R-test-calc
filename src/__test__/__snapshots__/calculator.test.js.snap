// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component Calculator match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="calculator-wrapper"
        data-id="calculator-wrapper"
      >
        <div
          class="calculator"
        >
          <div
            class="display"
          >
            <section
              class="calculator__presentation-box"
              data-testid="calculator-answer"
            />
            <section
              class="calculator__input-container"
            />
          </div>
          <section
            class="calculator__buttons-container"
          >
            <button
              class="calculator__button"
              data-testid="calculator-button-C"
              id="calculator-button-C"
              type="button"
            >
              <span>
                C
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-√"
              id="calculator-button-√"
              type="button"
            >
              <span>
                √
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-%"
              id="calculator-button-%"
              type="button"
            >
              <span>
                %
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-/"
              id="calculator-button-/"
              type="button"
            >
              <span>
                /
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-1"
              id="calculator-button-1"
              type="button"
            >
              <span>
                1
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-2"
              id="calculator-button-2"
              type="button"
            >
              <span>
                2
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-3"
              id="calculator-button-3"
              type="button"
            >
              <span>
                3
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-*"
              id="calculator-button-*"
              type="button"
            >
              <span>
                *
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-4"
              id="calculator-button-4"
              type="button"
            >
              <span>
                4
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-5"
              id="calculator-button-5"
              type="button"
            >
              <span>
                5
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-6"
              id="calculator-button-6"
              type="button"
            >
              <span>
                6
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button--"
              id="calculator-button--"
              type="button"
            >
              <span>
                -
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-7"
              id="calculator-button-7"
              type="button"
            >
              <span>
                7
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-8"
              id="calculator-button-8"
              type="button"
            >
              <span>
                8
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-9"
              id="calculator-button-9"
              type="button"
            >
              <span>
                9
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-+"
              id="calculator-button-+"
              type="button"
            >
              <span>
                +
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-("
              id="calculator-button-("
              type="button"
            >
              <span>
                (
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-)"
              id="calculator-button-)"
              type="button"
            >
              <span>
                )
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-0"
              id="calculator-button-0"
              type="button"
            >
              <span>
                0
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-."
              id="calculator-button-."
              type="button"
            >
              <span>
                .
              </span>
            </button>
            <button
              class="calculator__button"
              data-testid="calculator-button-="
              id="calculator-button-="
              type="button"
            >
              <span>
                =
              </span>
            </button>
          </section>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="calculator-wrapper"
      data-id="calculator-wrapper"
    >
      <div
        class="calculator"
      >
        <div
          class="display"
        >
          <section
            class="calculator__presentation-box"
            data-testid="calculator-answer"
          />
          <section
            class="calculator__input-container"
          />
        </div>
        <section
          class="calculator__buttons-container"
        >
          <button
            class="calculator__button"
            data-testid="calculator-button-C"
            id="calculator-button-C"
            type="button"
          >
            <span>
              C
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-√"
            id="calculator-button-√"
            type="button"
          >
            <span>
              √
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-%"
            id="calculator-button-%"
            type="button"
          >
            <span>
              %
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-/"
            id="calculator-button-/"
            type="button"
          >
            <span>
              /
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-1"
            id="calculator-button-1"
            type="button"
          >
            <span>
              1
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-2"
            id="calculator-button-2"
            type="button"
          >
            <span>
              2
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-3"
            id="calculator-button-3"
            type="button"
          >
            <span>
              3
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-*"
            id="calculator-button-*"
            type="button"
          >
            <span>
              *
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-4"
            id="calculator-button-4"
            type="button"
          >
            <span>
              4
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-5"
            id="calculator-button-5"
            type="button"
          >
            <span>
              5
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-6"
            id="calculator-button-6"
            type="button"
          >
            <span>
              6
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button--"
            id="calculator-button--"
            type="button"
          >
            <span>
              -
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-7"
            id="calculator-button-7"
            type="button"
          >
            <span>
              7
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-8"
            id="calculator-button-8"
            type="button"
          >
            <span>
              8
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-9"
            id="calculator-button-9"
            type="button"
          >
            <span>
              9
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-+"
            id="calculator-button-+"
            type="button"
          >
            <span>
              +
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-("
            id="calculator-button-("
            type="button"
          >
            <span>
              (
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-)"
            id="calculator-button-)"
            type="button"
          >
            <span>
              )
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-0"
            id="calculator-button-0"
            type="button"
          >
            <span>
              0
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-."
            id="calculator-button-."
            type="button"
          >
            <span>
              .
            </span>
          </button>
          <button
            class="calculator__button"
            data-testid="calculator-button-="
            id="calculator-button-="
            type="button"
          >
            <span>
              =
            </span>
          </button>
        </section>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
